src/svd/util.py

```diff
@@ -0,0 +1,386 @@
+
+    def build_bytes(self, selector: BuildSelector = BuildSelector()) -> bytearray:
+        """Encode device content as bytes.
+
+        :param selector: selected parts of the device.
+        :returns: content bytes.
+        """
+        memory = self.build_memory(selector)
+        out = bytearray()
+
+        for (addr_a, value_a), (addr_b, value_b) in pairwise(memory.items()):
+            if not out:
+                out.append(value_a)
+
+            num_empty = addr_b - addr_a - 1
+            if num_empty > 0:
+                # TODO: 0 may not always be valid
+                out.extend([0] * num_empty)
+
+            out.append(value_b)
+
+        return out
+
+    def build_memory(self, selector: BuildSelector = BuildSelector()) -> dict[int, int]:
+        """Encode device content as a mapping between address and value.
+
+        :param selector: selected parts of the device.
+        :returns: content memory map.
+        """
+        memory = {}
+
+        for peripheral in self._device.values():
+            if not selector.is_periph_selected(peripheral):
+                continue
+
+            if selector.content_status == BuildSelector.ContentStatus.MODIFIED:
+                # There's no good way to determine this in svada at the moment
+                periph_modified_filter = set()
+                for reg in peripheral.register_iter(leaf_only=True):
+                    if reg.modified:
+                        periph_modified_filter.update(reg.address_range)
+            else:
+                periph_modified_filter = None
+
+            memory_iter = peripheral.memory_iter(
+                absolute_addresses=True,
+                written_only=(
+                    selector.content_status == BuildSelector.ContentStatus.WRITTEN
+                ),
+            )
+
+            for addr, val in memory_iter:
+                if (
+                    periph_modified_filter is not None
+                    and addr not in periph_modified_filter
+                ):
+                    continue
+                memory[addr] = val
+
+        return memory
+
+    def build_dict(self, selector: BuildSelector = BuildSelector()) -> dict:
+        """Encode device content as a dictionary representation of the registers and content.
+
+        :param selector: selected parts of the device.
+        :returns: content dictionary.
+        """
+        config = {}
+
+        for peripheral in self._device.values():
+            if not selector.is_periph_selected(peripheral):
+                continue
+
+            cfg_periph = {}
+            nonleaf_stack = [cfg_periph]
+            for reg in peripheral.register_iter():
+                reg_depth = len(reg.path)
+                if reg_depth < len(nonleaf_stack):
+                    # We are returning from a nested context.
+                    # Prune any empty supertables on the way up.
+                    for elem in list(nonleaf_stack[reg_depth:]):
```

Comment:
There are a few minor bugs here, which lead to returning a non-empty dict even with no written registers:

1. This loop should start at `reg_depth - 1` to prune empty tables at the same level as the current register.
2. This loop should be reversed. The innermost empty tables must be pruned first before moving up.
3. This loop should be reused at the bottom. Pruning only from the top level is not enough, because if the last register has multiple segments in its path, it will leave a nested table.

Alternatively, I think the cleaner fix would be to just rewrite the function to be recursive.

Reply:
For reference, I found an easy way to build the dict without stacking *or* recursion:
```py
cfg_periph = {}

for reg in peripheral.register_iter(leaf_only=True):
    reg_table = cfg_periph
    for part in reg.path.parts:
        reg_table = reg_table.setdefault(str(part), {})

    for field_name, field in reg.fields.items():
        ... # Populate reg_table
```

Reply:
Ah, very nice, I'll integrate your solution. Should have probably done another pass on the implementation before posting the PR - thanks for the improvement.